network_wrangler/RoadwayNetwork.py

```diff
@@ -118,41 +118,57 @@ def write(self, filename: str, path: str = '.') -> None:
         shapes_file = os.path.join(path, filename + "_shape.geojson")
         self.shapes_df.to_file(shapes_file, driver='GeoJSON')
 
-
-
-    def apply_roadway_feauture_change(self, project_card: str) -> bool:
+    def apply_roadway_feauture_change(self, card_dict: dict) -> bool:
         '''
-        Changes the road network according to the project card passed
-        
+        Changes the road network according to the project card information passed
+
         args:
-        project_card: the path to the project card
-        
+        card_dict: dictionary with project card information
+
         returns:
         bool: True if successful.
         '''
-        #import pdb; pdb.set_trace()
-        with open(project_card, 'r') as card:
-            card_dict = yaml.safe_load(card)
-            
-            road_dict = card_dict.get("Road")            
-            name_str = road_dict.get(list(road_dict)[0])
-            id = name_str.split('=')[1]
-            
-            change_dict = card_dict.get("Change")
-            existing_value = change_dict.get(list(change_dict)[0])
-            build_value = change_dict.get(list(change_dict)[1])
-            
-            attribute = card_dict.get("Attribute")
-            
-            print(id)
-            print(attribute)
-            print(existing_value)
-            print(build_value)
-            
-            
-            #check existing value that it matches the attribute field in self.links_df
-            #TODO: how do i manipulate the data in the dataframe to apply the changes?
-            
-            self.links_df
-            
+
+        road_dict = card_dict.get("Road")
+        road_id = road_dict.get("Name").split("=")[1]
+
+        attribute = card_dict.get("Attribute").upper()
+        change_dict = card_dict.get("Change")
+        existing_value = change_dict.get("Existing")
+        build_value = change_dict.get("Build")
+
+        # identify the network link with same id as project card road id
+        # check if the attribute to be updated exists on the network links
+        # get the current attribute value for the link from the network
+        # check for current attribute value to be same as defined in the project card
+        # update the link attribute value
+
+        # TODO:
+        # change desired logger level for the different checks
+
+        found = False
+        for link in self.links_df['features']:
```

Comment:
Since this is pandas, you can avoid a for loop in general and do selections using the following notation:
`https://stackoverflow.com/questions/17071871/select-rows-from-a-dataframe-based-on-values-in-a-column-in-pandas`

Reply:
I would suggest creating a separate selection method (see my previous note with example) because you will need to select based on several dimensions and several ID types as well as stringing from A to B without a named street.

Reply:
> Since this is pandas, you can avoid a for loop in general and do selections using the following notation:
> `https://stackoverflow.com/questions/17071871/select-rows-from-a-dataframe-based-on-values-in-a-column-in-pandas`

The dataframe in links_df is two columns - `type` and `features`. links_df['features'] is a list of dictionaries, I don't think we can apply selection using the pandas select method. 
Instead of for loop, I can do something like below to create a selection

`next(link for link in net.links_df['features'] if link['osmid'] == '18218485')`
OR
`list(filter(lambda link: link['osmid'] == '18218485', net.links_df['features']))`
`

Reply:
Good point. I forgot that it was nested (yay json). Do you think we should normalize it (i.e. https://www.kaggle.com/jboysen/quick-tutorial-flatten-nested-json-in-pandas) or leave it as-is?

Reply:
yes, we could normalize it to make it easy to work with the links dataframe. But if do it in `Read` method and the links_df is flat, would be able to write it out back as nested json in `write` (I assume we want `write` method to output json file)?

Reply:
Yes, you'd also  need to update the existing write method.

Reply:
What I was saying is that if we normalize the link data (in `Read` function) to flat everything (including `locationReferences` nest), we wouldn't be able to write it back as deep nested json object in `write`. What do you think?